src/models/users.js

```diff
@@ -9,22 +6,65 @@ import {
 
 // Supported authentication methods.
 const authMethods = {
-  BASIC: 'basic'
+  BASIC: ({ username, password }) => {
+    // For now we only support admin authentication.
+    if (username !== 'admin' || password !== config.get('adminPass')) {
+      return Promise.reject(new Error(UNAUTHORIZED));
+    }
+
+    return Promise.resolve({
+      id: 'admin',
+      scope: 'admin'
+    });
+  },
+
+  AUTH_PROVIDER: (data) => {
+    return Promise.resolve({
+      id: data,
+      scope: 'user'
+    });
+  },
+};
+
+module.exports = (sequelize, DataTypes) => {
+  const User = sequelize.define('Users', {
+    opaqueId: { type: DataTypes.STRING(256) },
+    provider: { type: DataTypes.STRING(32) },
+  }, {
+    indexes: [{
+      unique: true,
+      fields: ['opaqueId', 'provider', 'ClientKey']
+    }]
+  });
+
+  User.authenticate = (method, data) => {
+    if (!authMethods[method]) {
+      return Promise.reject(new Error(UNSUPPORTED_AUTH_METHOD));
+    }
+
+    return authMethods[method](data)
+      .then(userData => {
+        if (userData.scope !== 'user') {
+          return userData;
+        }
+
+        return User.findOrCreate({
+          attributes: [],
+          where: userData.id, // this contains all user attributes
+        })
+          .then(() => userData);
```

Comment:
Note that the result to `authenticate` is not returned to the user, rather it's stored in `req.user` for the next (express) middlewares.

Currently I'm not returning the sequelize's user at all. `userData` comes from the authentication methods. Depending on the methods we don't always have the same structure. My take at this is that the "scope" will give us the structure.

* `userData.scope === 'user'` => `userData.id` is an object with `opaqueId`, `provider` and `ClientKey`. This looks like the sequelize object but it's not. It comes from the authentication methods to we rely on them giving us the right structure (but a wrong structure would make the find fail, so I think we're good).

  My main issue with the current approach is that in this function we never know explicitely what's inside `userData.id`, that could make it more difficult to read the function.

* `userData.scope === 'admin'` => `userData.id` is `'admin'`
* `userData.scope === 'client'` => `userData.id` is just the client's key

So once we know the `scope` we also know how to consult the `id`.